### PR TITLE
Fix minimization to be insensitive to redundant graph constraints

### PR DESCRIPTION
**Pull request by mattam82**
This makes hott compile with a few modifications, performance should be evaluated too.
Currently being tested on the contribs here: https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/441/

### CODE REVIEW

**Comment by mattam82** [comment]
Remove approximation in checking if a universe is lower than existing ones. https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/480/

**Comment by mattam82** [comment]
The refined patch passed the contribs. I'd be happier with a review before merging though. https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/482/

File: library/universes.ml [hunk 1]
```diff
@@ -702,12 +702,43 @@ let pr_universe_body = function
...
 let pr_universe_opt_subst = Univ.LMap.pr pr_universe_body
 
-exception Found of Level.t
+let compare_constraint_type d d' =
+  match d, d' with
+  | Eq, Eq | Le, Le | Lt, Lt -> 0
+  | Eq, _ -> -1
+  | Le, Eq -> 1
+  | Le, Lt -> -1
+  | Lt, (Eq | Le) -> 1
```
Comment (silene) [inline library/universes.ml]: Please make it so that the ordering is obvious. I don't know if it is the most efficient one, but it could be as follows:

    | Eq, Eq -> 0
    | Eq, _ -> -1
    | _, Eq -> 1
    | Le, Le -> 0
    | Le, _ -> -1
    | _, Le -> 1
    | Lt, Lt -> 0

File: library/universes.mli [hunk 2]
```diff
@@ -243,28 +243,31 @@ val choose_canonical : universe_set -> (Level.t -> bool) (* flexibles *) -> univ
...
   Univ.LSet.t * Univ.universe option Univ.LMap.t *
     Univ.LSet.t *
-    (bool * bool * Univ.universe) Univ.LMap.t * Univ.constraints ->
+    (bool * bool * Univ.universe * lowermap) Univ.LMap.t * Univ.constraints ->
   (Univ.LSet.t * Univ.universe option Univ.LMap.t *
     Univ.LSet.t *
-     (bool * bool * Univ.universe) Univ.LMap.t * Univ.constraints) *
-    (bool * bool * Univ.universe)
+     (bool * bool * Univ.universe * lowermap) Univ.LMap.t * Univ.constraints) *
+    (bool * bool * Univ.universe * lowermap)
```
Comment (silene) [inline library/universes.mli]: Please give a type name to these quadruplets. It would be even better if you could turn them into records.

File: library/universes.ml [hunk 3]
```diff
@@ -756,54 +793,78 @@ let minimize_univ_variables ctx us algs left right cstrs =
...
-      with Not_found -> acc, []
+	  let acc, left, newlow, lower =
+            List.fold_left
+	      (fun (acc, left', newlow, lower') (d, l) ->
+	        let acc', (enf,alg,l',lower) = aux acc l in
+	        let l' =
+		  if enf then Universe.make l
+		  else l'
+	        in acc', (d, l') :: left',
+                   lower_add l d newlow, lower_union lower lower')
```
Comment (silene) [inline library/universes.ml]: Could there be a conflict when doing the union here? (I am always wary of map unions.)

File: library/universes.mli [hunk 4]
```diff
@@ -243,28 +243,31 @@ val choose_canonical : universe_set -> (Level.t -> bool) (* flexibles *) -> univ
     
 val compute_lbound : (constraint_type * Univ.universe) list -> universe option
 
+type lowermap = constraint_type Univ.LMap.t
+
 val instantiate_with_lbound :
```
Comment (silene) [inline library/universes.mli]: Does it make sense to actually export this function?

File: library/universes.ml [hunk 5]
```diff
@@ -756,54 +793,78 @@ let minimize_univ_variables ctx us algs left right cstrs =
...
+                let d =
+                  if i == 0 then d
+                  else match d with
+                       | Le -> Lt
+                       | d -> d
+                in
+                try let d' = LMap.find l lower in
+                    (* If d is stronger than the already implied lower
+                     * constraints we must keep it *)
+                    compare_constraint_type d d' > 0
```
Comment (silene) [inline library/universes.ml]: The logic is a bit hard to follow here. I have no way of telling if this part works correctly.

File: library/universes.mli [hunk 6]
```diff
@@ -243,28 +243,31 @@ val choose_canonical : universe_set -> (Level.t -> bool) (* flexibles *) -> univ
...
-    (bool * bool * Univ.universe) Univ.LMap.t * Univ.constraints ->
+    (bool * bool * Univ.universe * lowermap) Univ.LMap.t * Univ.constraints ->
   (Univ.LSet.t * Univ.universe option Univ.LMap.t *
     Univ.LSet.t *
-     (bool * bool * Univ.universe) Univ.LMap.t * Univ.constraints) *
-    (bool * bool * Univ.universe)
+     (bool * bool * Univ.universe * lowermap) Univ.LMap.t * Univ.constraints) *
+    (bool * bool * Univ.universe * lowermap)
 
 val minimize_univ_variables :
```
Comment (maximedenes) [inline library/universes.mli]: How do you expect anyone to make sense of this type?

File: library/universes.mli (hunk 2 above)
Comment (maximedenes) [inline library/universes.mli]: Yes, stop with these tuples, pleeease! We have enough to fix in the existing code.

Comment (mattam82) [inline library/universes.mli]: That shouldn't be exported either.

File: library/universes.ml (hunk 5 above)
Comment (mattam82) [inline library/universes.ml]: It's unclean I agree, I had to hack around the interface in Univ.Universe. I'm not sure how to help here, I just stared at it long enough. We're checking if (d,l) is already implied by the lower constraints on some level u. If it represents l < u (d is Lt or d is Le and i > 0, the i < 0 case is impossible due to invariants of Univ), and the lower constraints only have l <= u then it is not implied.

Comment (mattam82) [inline library/universes.ml]: I added the comment to the code

**Comment by mattam82** [comment]
Cleaned and rebased, this should be merged once this check goes through:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/523/

**Comment by mattam82** [comment]
Trying again, the cleanup made compilation fail 
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/535/console
